src/node_url_pattern.h

```diff
@@ -0,0 +1,112 @@
+};
+
+class URLPattern : public BaseObject {
+ public:
+  URLPattern(Environment* env,
+             v8::Local<v8::Object> object,
+             ada::url_pattern<URLPatternRegexProvider>&& url_pattern);
+
+  static void New(const v8::FunctionCallbackInfo<v8::Value>& args);
+
+  // V8 APIs
+  // - Functions
+  static void Exec(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Test(const v8::FunctionCallbackInfo<v8::Value>& info);
+  // - Getters
+  static void Hash(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Hostname(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Password(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Pathname(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Port(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Protocol(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Search(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void Username(const v8::FunctionCallbackInfo<v8::Value>& info);
+  static void HasRegexpGroups(const v8::FunctionCallbackInfo<v8::Value>& info);
+
+  void MemoryInfo(MemoryTracker* tracker) const override;
+  SET_MEMORY_INFO_NAME(URLPattern)
+  SET_SELF_SIZE(URLPattern)
+
+  class URLPatternInit {
+   public:
+    static ada::url_pattern_init FromJsObject(Environment* env,
+                                              v8::Local<v8::Object> obj);
+    static v8::MaybeLocal<v8::Value> ToJsObject(
+        Environment* env, const ada::url_pattern_init& init);
+  };
+
+  class URLPatternOptions {
+   public:
+    static std::optional<ada::url_pattern_options> FromJsObject(
+        Environment* env, v8::Local<v8::Object> obj);
+  };
+
+  class URLPatternResult {
+   public:
+    static v8::MaybeLocal<v8::Value> ToJSValue(
+        Environment* env, const ada::url_pattern_result& result);
+  };
+
+  class URLPatternComponentResult {
+   public:
+    static v8::MaybeLocal<v8::Object> ToJSObject(
+        Environment* env, const ada::url_pattern_component_result& result);
+  };
+
+ private:
+  ada::url_pattern<URLPatternRegexProvider> url_pattern_;
+  // Getter methods
+  v8::MaybeLocal<v8::Value> Hash() const;
+  v8::MaybeLocal<v8::Value> Hostname() const;
+  v8::MaybeLocal<v8::Value> Password() const;
+  v8::MaybeLocal<v8::Value> Pathname() const;
+  v8::MaybeLocal<v8::Value> Port() const;
+  v8::MaybeLocal<v8::Value> Protocol() const;
+  v8::MaybeLocal<v8::Value> Search() const;
+  v8::MaybeLocal<v8::Value> Username() const;
+  bool HasRegExpGroups() const;
+  // Public API
+  v8::MaybeLocal<v8::Value> Exec(
+      Environment* env,
+      const ada::url_pattern_input& input,
+      std::optional<std::string_view>& baseURL);  // NOLINT (runtime/references)
+  bool Test(
+      Environment* env,
+      const ada::url_pattern_input& input,
+      std::optional<std::string_view>& baseURL);  // NOLINT (runtime/references)
```

Comment:
Any reason to prefer NOLINT over the recommendation here?

(Although I feel like in this case this maybe should be a const reference anyway?)

Reply:
If I make it a const parameter, then it doesn't compile due to this line:

```cpp
if (auto result = url_pattern_.exec(input, baseURL ? &*baseURL : nullptr)) {
```

exec() method in ada expects `std::string_view* base_url` where ada::url has a similar API surface as well, and making it const would differentiate the signature between url_pattern and url in Ada.

I'm not sure what is the best way, other than adding NOLINT and keeping this non-const.

Reply:
blech... but yeah, that's better even if it looks ugly ;-)